### PR TITLE
Full Code Review

### PR DESCRIPTION
**Pull request by dilzio**
This PR captures code review comments for the codebase.

### CODE REVIEW

File: indexer/indexer.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+	DB storage.Storage = storage.NewLocalMapStorage()
+)
+
+// wikipediafetcher is a wrapper fetcher function for the Wikipedia Pageviews API.
+func wikipediafetcher(date time.Time) ([]messages.ArticleCount, error) {
+	counts := []messages.ArticleCount{}
+	year := strconv.Itoa(date.Year())
+	month := date.Format(constants.TWODAYMONTH)
+	day := date.Format(constants.TWODAYDAYOFWEEK)
+	url := fmt.Sprintf(constants.PAGEVIEWS_URL, year, month, day)
```
Comment (dilzio) [inline indexer/indexer.go:39]: consider fmt.Errorf() instead to allow for more descriptive error messages

File: indexer/indexer.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	index := sortedset.New[string, int, messages.ArticleCount]()
+	ssUpdateMutex := sync.Mutex{}
+	errorChannel := make(chan error, constants.MAXDAYINTERVAL)
+	for d := startdate; !d.After(enddate) == true; d = d.AddDate(0, 0, 1) {
+		wg.Add(1)
+		go func(date time.Time) {
+			defer wg.Done()
+			countsForDay, err := getArticleCountsForDay(date)
+			if err != nil {
+				log.Debugf("Unable to retrieve data for date: %v", date)
```
Comment (dilzio) [inline indexer/indexer.go:131]: add error logging here

File: indexer/indexer.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+// Function GetCountsForArticleInRange assembles a total view count for q specific article in a date range
+func GetCountsForArticleInRange(article string, startdate time.Time, enddate time.Time) (messages.ArticleCountsForDateRange, error) {
+	wg := sync.WaitGroup{}
+	index := sortedset.New[string, int, messages.ArticleCount]()
+	ssUpdateMutex := sync.Mutex{}
+	errorChannel := make(chan error, constants.MAXDAYINTERVAL)
+	for d := startdate; !d.After(enddate) == true; d = d.AddDate(0, 0, 1) {
+		wg.Add(1)
+		go func(date time.Time) {
```
Comment (dilzio) [inline indexer/indexer.go:127]: consider refactoring this to a named function for readability

File: indexer/indexer.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
+// Package indexer contains functions for sourcing and aggregating article counts. It makes heavy use of the
+// github.com/zavitax/sortedset-go implementation for ranking.  It depends on a Storage implementation for caching
+// and a fetcher function for retrieving the articles from the original source
+// TODO: these api calls have a mostly similar algorithm that could be refactored into a higher-order function
```
Comment (dilzio) [inline indexer/indexer.go:4]: Per the TODO...look into opportunities to refactor

File: constants/constants.go [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package constants
+
+// see time package docs for why go formats days like this
+const DATELAYOUT = "20060102"
```
Comment (dilzio) [inline constants/constants.go:4]: consider bracket form for declarations

File: storage/storage.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		sync.RWMutex{},
+	}
+}
+
+// Add an article day count
+func (t *LocalMapStorage) Put(key time.Time, value []messages.ArticleCount) {
+	key = key.Truncate(TRUNCATE_TO_DAY)
+	t.rwMutex.Lock()
+	defer t.rwMutex.Unlock()
+	t.internal[key] = value
```
Comment (dilzio) [inline storage/storage.go:40]: eating a potential error here....should catch and handle

File: storage/storage.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+func NewLocalMapStorage() *LocalMapStorage {
+	return &LocalMapStorage{
+		make(map[time.Time][]messages.ArticleCount),
+		sync.RWMutex{},
+	}
+}
+
+// Add an article day count
+func (t *LocalMapStorage) Put(key time.Time, value []messages.ArticleCount) {
+	key = key.Truncate(TRUNCATE_TO_DAY)
```
Comment (dilzio) [inline storage/storage.go:37]: it's better to let the client do the truncation as they may insights into key creation not apparent here.

File: storage/storage.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	t.rwMutex.Lock()
+	defer t.rwMutex.Unlock()
+	t.internal[key] = value
+}
+
+// Retrieve a pointer to an article day count. Second return value will be true if the key is present
+func (t *LocalMapStorage) Get(key time.Time) ([]messages.ArticleCount, bool) {
+	key = key.Truncate(TRUNCATE_TO_DAY)
```
Comment (dilzio) [inline storage/storage.go:45]: same comment as for Put function.

File: main/main.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/go-chi/chi/v5/middleware"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"pelotechfun/service"
+)
+
+func main() {
+	// Set up OpenTelemetry.
+	otelShutdown, err := setupOTelSDK(context.Background())
+	if err != nil {
```
Comment (dilzio) [inline main/main.go:16]: Use a more descriptive error message and kill if otel fails to initialize properly.

File: main/main.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	log "github.com/sirupsen/logrus"
```
Comment (dilzio) [inline main/main.go:8]: Still have two logging frameworks - should  complete migration to otel

File: main/main.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+		err = errors.Join(err, otelShutdown(context.Background()))
+	}()
+	log.SetLevel(log.InfoLevel)
+	r := chi.NewRouter()
+	r.Use(middleware.Logger)
+	r.Get("/mostviewed/{startdate}/{enddate}", service.DoGetArticleCountsForDateRange)
+	r.Get("/viewcount/{article}/{startdate}/{enddate}", service.DoCalcViewCountForArticle)
+	r.Get("/mostviewedday/{article}/{year}/{month}", service.DoCalcMostViewedDayInMonthForArticle)
+	log.Infof("Hi! listening on localhost:8080")
+	log.Fatal(http.ListenAndServe(":8080", r))
```
Comment (dilzio) [inline main/main.go:30]: consider adding a log message indicating successful startup

File: go.mod [hunk 12]
```diff
@@ -0,0 +1,30 @@
+module pelotechfun
+
+go 1.22.3
+
+require (
+	github.com/go-chi/chi/v5 v5.0.12
+	github.com/sirupsen/logrus v1.9.3
```
Comment (dilzio) [inline go.mod:7]: remove logrus and complete otel migration

File: service/article_stats_service.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+		return
+	}
+	var bytes []byte
+	if bytes, err = json.Marshal(&result); err != nil {
+		log.Error("Failed to marshal reply:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(bytes)
```
Comment (dilzio) [inline service/article_stats_service.go:59]: this code can be refactored into a separate function and shared with DoGetArticleCountsForDateRange()

File: main/app_verifier_test.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+	/*
+		Example code to show how Wikipedia fetcher can be stubbed out:
+		if (config.usestub){
+			indexer.Fetcher = func(date time.Time) ([]messages.ArticleCount, error) {
+				// returns stubbed data
+			}
+		}
+	*/
+	go main()
+	time.Sleep(4 * time.Second)
```
Comment (dilzio) [inline main/app_verifier_test.go:28]: Sleeping could result in flaky test behavior...better to find a deterministic way of knowing when the server is up

File: main/app_verifier_test.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+	//mostviewed HappyPath
+	r, _ := http.Get("http://localhost:8080/mostviewed/20220101/20220102")
+	defer r.Body.Close()
+	bytes, _ := io.ReadAll(r.Body)
+	payloadString := string(bytes[:])
+	assert.True(t, strings.Contains(payloadString, "\"Main_Page\",\"views\":10226718"))
```
Comment (dilzio) [inline main/app_verifier_test.go:34]: Can we improve DRY here by removing cut and paste?  Maybe candidate for table driven tests

File: main/app_verifier_test.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+		if (config.usestub){
+			indexer.Fetcher = func(date time.Time) ([]messages.ArticleCount, error) {
+				// returns stubbed data
+			}
+		}
+	*/
+	go main()
+	time.Sleep(4 * time.Second)
+	//mostviewed HappyPath
+	r, _ := http.Get("http://localhost:8080/mostviewed/20220101/20220102")
```
Comment (dilzio) [inline main/app_verifier_test.go:30]: error not handled

File: main/app_verifier_test.go [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+	defer r.Body.Close()
+	bytes, _ := io.ReadAll(r.Body)
```
Comment (dilzio) [inline main/app_verifier_test.go:32]: handle error

File: main/app_verifier_test.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	payloadString := string(bytes[:])
+	assert.True(t, strings.Contains(payloadString, "\"Main_Page\",\"views\":10226718"))
+
+	//mostviewed bad dates
+	r, _ = http.Get("http://localhost:8080/mostviewed/20220101/")
+	defer r.Body.Close()
```
Comment (dilzio) [inline main/app_verifier_test.go:38]: more idiomatic to write:

r, err := http.Get("http://localhost:8080/mostviewed/20220101/20220102")
if err != nil {
	assert.Fail(t, "Failed making request")
	return
}
defer r.Body.Close()

File: indexer/indexer.go [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+	//Call the API and return an error if any
+	resp, err := http.Get(url)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		message := "Unable to retrieve page count data from Wikipedia: " + date.Format(constants.DATELAYOUT)
+		log.Error(message)
+		return counts, errors.New(message)
+	}
+	//Map body into struct representation, return an error if it fails
+	defer resp.Body.Close()
+	body, _ := io.ReadAll(resp.Body)
```
Comment (dilzio) [inline indexer/indexer.go:50]: handle error here.

File: indexer/indexer.go (hunk 1 above)
Comment (dilzio) [inline indexer/indexer.go:39]: Fetcher should be moved to its own package as the feature is distinct from the indexing functions below.

From a design point of view we should consider moving wikipedia fetching and caching functionality to its own service.  This will allow us to scale indexing with less impact on wikipedia (i.e. running up against rate limiting) as well as being able to share cached records among indexing instances.

**Review by dilzio — COMMENTED** [review]
Moving this codebase from the prototype example to something production worthy would require finishing the implementation:

- Integrating an appropriate Storage implementation
- Refactor Wikipedia fetching into another service (see indexer.go)
- Remove legacy logging framework (logrus) and finish up Open Telemetry impl. 
- Expand overall logging
- Implement client solution for handling Wikipedia rate limiting
- Externalize config
- various improvements to error handling and uses of idiomatic Go (captured in the comments)
- More tests including fuzz testing to check for boundary conditions
